File: Top200/25-ReverseKGroup.java
package Top200;

class ReverseKGroup {
    class ListNode {
        int val;
        ListNode next;

        ListNode (int val) {
            this.val = val;
        }
    }

    public ListNode reverseKGroup (ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }

        if (k == 1) {
            return head;
        }

        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        prev.next = head;

        while (head != null) {
            ListNode tail = prev;

            for (int i = 0; i < k; i++) {
                tail = tail.next;

                if (tail == null) {
                    return dummy.next;
                }
            }

            ListNode tailNext = tail.next;
            tail.next = null;

            ListNode prevIn = null;
            ListNode currIn = head;
            while (currIn != null) {
                ListNode temp = currIn.next;
                currIn.next = prevIn;
                prevIn = currIn;
                currIn = temp;
            }

            prev.next = prevIn;
            head.next = tailNext;

            prev = head;
            head = tailNext;
        }

        return dummy.next;
    }
}

File: Top200/25-ReverseKGroup.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	if k == 1 {
		return head
	}

	// 主要思路就是：
	// 1. 先找到 k 个节点的区间，然后对这个区间内的链表进行翻转；
	// 2. 按照这种方式完成整个链表的翻转
	dummy := &ListNode{Val: -1}
	prev := dummy
	prev.Next = head

	for head != nil {
		// 先找到 k 个节点的区间
		tail := prev
		for i := 0; i < k; i++ {
			tail = tail.Next

			if tail == nil {
				return dummy.Next
			}
		}

		tailNext := tail.Next
		tail.Next = nil

		var prevIn *ListNode = nil
		var currIn *ListNode = head

		for currIn != nil {
			temp := currIn.Next
			currIn.Next = prevIn
			prevIn = currIn
			currIn = temp
		}

		prev.Next.Next = tailNext
		prev.Next = tail

		prev = head
		head = tailNext
	}

	return dummy.Next
}
